include/uart.h

```diff
@@ -116,8 +229,19 @@ struct uart_device_config {
 #endif
 };
 
+#endif /* CONFIG_UART_HIGH_LEVEL_API */
+
 /** @brief Driver API structure. */
 struct uart_driver_api {
+
+#ifdef CONFIG_UART_HIGH_LEVEL_API
+
+	int (*xfer)(struct device *dev, char *buf, size_t len, int dir);
```

Comment:
Full-duplex transfers can be handled in non-blocking mode, by independently setting buffers using uart_rx and uart_tx functions.

I haven't seen UART drivers that handle full-duplex transmission in blocking mode, so I don't think that function like spi_transceive for setting both buffers at the same time is needed. If there is a use case then I can change that.

Reply:
> Full-duplex transfers can be handled in non-blocking mode, by independently setting buffers using uart_rx and uart_tx functions.

The problem here is the driver will not know when the actual transfer should start.

Reply:
TX transfer should start immediately after uart_tx function is called.
RX transfer should start when data is available on line.

Reply:
So you might have timing issue then? Let's say you start tx, but the time you start rx, something already came in and your driver just dropped it. Does that sound plausible or am I being on a far fetched assumption?

Reply:
> but the time you start rx, something already came in and your driver just dropped it.

Not directly related to this sub-thread, but on a more general level, I have the same concern as you, @tbursztyka - this API doesn't seem to be high-level enough and still leaves too many underspecified and "can go wrong" moments.

I guess the root problem is that this tries to do it "like SPI or I2C", but there's one big difference between those and UART. SPI and I2C are synchronous, master-slave protocols. But UART is async "peer to peer" protocol. Even, with optional peer to peer flow control. This concept of "transfers" is poorly suited for it. "Always available buffers" is better notion, but then there's already https://github.com/zephyrproject-rtos/zephyr/pull/10765 .

Reply:
I've had a following concept in mind for this API:
For TX transfers, there is no risk of losing data, so it should offer a simple function uart_tx that accepts pointer to a buffer and length to be transmitted. After transmission is finished it can be called again to transmit another chunk of data.

For RX transfers, there has to be a way to receive data without losing it. So at start we should set up two buffers using uart_rx function. This way when first buffer is filled, driver starts receiving to second buffer and calls event handler. In event handler user provides another buffer for receiving by calliing uart_rx again and operates on data stored in first buffer.
This way transmission continues alternating between two buffers and giving time for application to process incoming data.

Reply:
Sounds reasonable to me

Reply:
Have you read the serial port driver on arduino? I think that's is a good way, every hardware has a rx ringbuffer for buffered all the received data in the background. The high level APIs don't need deal with the interrupts, they can read data from rx ringbuffer directly.

Also there can has a tx ringbuffer if needed, buffered all datas the high level to send, sent them when the bus is not busy.

In my opinion we can implement the high level based on top of uart interrupt, like whats modem receiver did.

Reply:
@qianfan-Zhao : That's what subsys/console has been doing for 2nd year or so, with API being gradually refactored towards POSIX-likeness, the latest piece is at https://github.com/zephyrproject-rtos/zephyr/pull/10765 . It goes slowly because every Zephyr developer wants their own original API, that's how we got original UART line input, uart_pipe, more recently modem_receiver, and now this API.

Reply:
Not every developer wants to implement their own application interface, but there isn't enough good design yet, because UART is so special relative to SPI or I2C.

Reply:
> but there isn't enough good design yet

Yes, that's why we all are here, to make sure that we don't repeat previous mistakes. There're well-known best practices how to deal with this stuff, both algorithmically and API-wise, we just need to accept them and implement.

Reply:
let me put my 2 cents to this. i've been using following uart rx approach for some time and that worked well (contrary to many other in the past):
- uart is initialized with rx timeout
- 2 rx event types: RX_BUF_REQUEST, RX_RDY(data, len)
- rx_buf_rsp(buf, len) function for providing rx buf
- rx_enable(buf, len) - function for starting RX with initial buffer.

RX_RDY is generated when timeout occurs or buffer is fully filled. In that case RX_BUF_REQUEST is also triggered.
RX is started with first buffer provided. RX_BUF_REQUEST will be immediately triggered to get next buffer. Ideally, driver shall operate in a way that it has always one buffer ahead. If HW supports DMA linked transfers it will be possible to ensure that there is no gap in reception (even reliable reception without flow control).

Upper layer (hw independent thus common, can be console) has pool of buffers (min 3) or ring buffer.  On RX_RDY can copy that to ring_buffer (or it can even use data directly from ring_buffer using recently added api) and takes care of managing memory. In simplest case (no dma) buffers are 1 byte long and timeout is set to 0. In order to make the timeout work with dma, uart must support hw timeouting (nrf uarte supports that, i've seen also hw support for rx timeout in silabs efr32).

@Mierunski  proposed something similar:
> So at start we should set up two buffers using uart_rx function

But i think that explicit buf_req and buf_rsp is cleaner.